File: app/src/main/java/com/idormy/sms/forwarder/utils/BatteryUtils.java
package com.idormy.sms.forwarder.utils;

import android.annotation.SuppressLint;
import android.content.Intent;
import android.os.BatteryManager;
import android.util.Log;

public class BatteryUtils {
    private static final String TAG = "BatteryUtils";

    @SuppressLint("DefaultLocale")
    public static String getBatteryInfo(Intent intent) {
        Log.i(TAG, "getBatteryInfo--------------");
        String action = intent.getAction();
        Log.i(TAG, " 0 action:" + action);
        Log.i(TAG, "ACTION_BATTERY_CHANGED");
        int status = intent.getIntExtra("status", 0);
        int health = intent.getIntExtra("health", 0);
        //boolean present = intent.getBooleanExtra("present", false);
        int levelCur = intent.getIntExtra("level", 0);
        int scale = intent.getIntExtra("scale", 0);
        //int icon_small = intent.getIntExtra("icon-small", 0);
        int plugged = intent.getIntExtra("plugged", 0);
        int voltage = intent.getIntExtra("voltage", 0);
        int temperature = intent.getIntExtra("temperature", 0);
        //String technology = intent.getStringExtra("technology");

        String msg = "";
        msg += "\n剩余电量：" + levelCur + "%";

        if (scale > 0) msg += "\n充满电量：" + scale + "%";

        if (voltage > 0) msg += "\n当前电压：" + String.format("%.2f", voltage / 1000F) + "V";

        if (temperature > 0) msg += "\n当前温度：" + String.format("%.2f", temperature / 10F) + "℃";

        msg += "\n电池状态：" + getStatus(status);

        if (health > 0) msg += "\n健康度：" + getHealth(health);

        switch (plugged) {
            case BatteryManager.BATTERY_PLUGGED_AC:
                msg += "\n充电器：AC";
                break;
            case BatteryManager.BATTERY_PLUGGED_USB:
                msg += "\n充电器：USB";
                break;
            case BatteryManager.BATTERY_PLUGGED_WIRELESS:
                msg += "\n充电器：无线";
                break;
        }

        Log.i(TAG, msg);
        return msg;
    }

    //电池状态
    public static String getStatus(int status) {
        switch (status) {
            case BatteryManager.BATTERY_STATUS_CHARGING:
                return "充电中";
            case BatteryManager.BATTERY_STATUS_DISCHARGING:
                return "放电中";
            case BatteryManager.BATTERY_STATUS_NOT_CHARGING:
                return "未充电";
            case BatteryManager.BATTERY_STATUS_FULL:
                return "充满电";
            case BatteryManager.BATTERY_STATUS_UNKNOWN:
            default:
                return "未知";
        }
    }

    //健康度
    public static String getHealth(int health) {
        switch (health) {
            case 2:
                return "良好";
            case 3:
                return "过热";
            case 4:
                return "没电";
            case 5:
                return "过电压";
            case 6:
                return "未知错误";
            case 7:
                return "温度过低";
            default:
            case 1:
                return "未知";
        }
    }
}

File: app/src/main/java/com/idormy/sms/forwarder/utils/BatteryUtils.kt
package com.idormy.sms.forwarder.utils

import android.content.Intent
import android.os.BatteryManager
import com.idormy.sms.forwarder.R
import com.idormy.sms.forwarder.entity.BatteryInfo
import com.xuexiang.xutil.resource.ResUtils.getString

@Suppress("MemberVisibilityCanBePrivate")
object BatteryUtils {
    //private const val TAG = "BatteryUtils"

    fun getBatteryInfo(intent: Intent?): BatteryInfo {
        val batteryInfo = BatteryInfo()
        if (intent == null) return batteryInfo

        val level = intent.getIntExtra("level", 0)
        val scale = intent.getIntExtra("scale", 0)
        val voltage = intent.getIntExtra("voltage", 0)
        val temperature = intent.getIntExtra("temperature", 0)
        val status = intent.getIntExtra("status", 0)
        val health = intent.getIntExtra("health", 0)
        val plugged = intent.getIntExtra("plugged", 0)

        batteryInfo.level = "$level%"
        if (scale > 0) batteryInfo.scale = "$scale%"
        if (voltage > 0) batteryInfo.voltage = "${String.format("%.2f", voltage / 1000f)}V"
        if (temperature > 0) batteryInfo.temperature = "${String.format("%.2f", temperature / 10f)}℃"
        batteryInfo.status = getStatus(status)
        batteryInfo.health = getHealth(health)
        batteryInfo.plugged = getPlugged(plugged)
        //Log.i(TAG, batteryInfo.toString())
        return batteryInfo
    }

    fun getStatus(status: Int): String {
        return when (status) {
            BatteryManager.BATTERY_STATUS_CHARGING -> getString(R.string.battery_charging)
            BatteryManager.BATTERY_STATUS_DISCHARGING -> getString(R.string.battery_discharging)
            BatteryManager.BATTERY_STATUS_NOT_CHARGING -> getString(R.string.battery_not_charging)
            BatteryManager.BATTERY_STATUS_FULL -> getString(R.string.battery_full)
            BatteryManager.BATTERY_STATUS_UNKNOWN -> getString(R.string.battery_unknown)
            else -> getString(R.string.battery_unknown)
        }
    }

    fun getHealth(health: Int): String {
        return when (health) {
            BatteryManager.BATTERY_HEALTH_UNKNOWN -> getString(R.string.battery_unknown)
            BatteryManager.BATTERY_HEALTH_GOOD -> getString(R.string.battery_good)
            BatteryManager.BATTERY_HEALTH_OVERHEAT -> getString(R.string.battery_overheat)
            BatteryManager.BATTERY_HEALTH_DEAD -> getString(R.string.battery_dead)
            BatteryManager.BATTERY_HEALTH_OVER_VOLTAGE -> getString(R.string.battery_over_voltage)
            BatteryManager.BATTERY_HEALTH_UNSPECIFIED_FAILURE -> getString(R.string.battery_unspecified_failure)
            BatteryManager.BATTERY_HEALTH_COLD -> getString(R.string.battery_cold)
            else -> getString(R.string.battery_unknown)
        }
    }

    fun getPlugged(plugged: Int): String {
        return when (plugged) {
            BatteryManager.BATTERY_PLUGGED_AC -> getString(R.string.battery_ac)
            BatteryManager.BATTERY_PLUGGED_USB -> getString(R.string.battery_usb)
            BatteryManager.BATTERY_PLUGGED_WIRELESS -> getString(R.string.battery_wireless)
            else -> getString(R.string.battery_unknown)
        }
    }

}